src/spikeinterface/core/base.py

```diff
@@ -164,7 +164,7 @@ def id_to_index(self, id) -> int:
     def annotate(self, **new_annotations) -> None:
         self._annotations.update(new_annotations)
 
-    def set_annotation(self, annotation_key, value: Any, overwrite=False) -> None:
+    def set_annotation(self, annotation_key: str, value: Any, overwrite=False) -> None:
```

Comment:
So value could be like a lamda function? like

```python
set_annotation(annotation_key = 'my_lambda', value = lamda x)
```

that might be a little pendantic, but it really just comes back to our debate about explicit vs too much info.

Reply:
Ah, but I just remember you don't think ArrayLike is very useful, right? Probably less so in this case where you can pass ints, floats and strings that are array like but not tuple, liste or np.ndarray.

I am fine with the way it is now. I think that Aurelien added this so his linter would not look ugly : P

What would you have here? Any suggestions?

Reply:
I'm fine with how it is now. I guess for this it was more supporting Sam in that sometimes typing in python is a mess :) 

But for now I think `Any` is fine.

Reply:
I disagree with this assesment. What's the mess here? It is just hard to type but it is also hard to code. We really have not set those restrictions in place any way.

If anything it reveals that we have not constrain our schema of annotations enough.

Reply:
To be fair I never use annotations, so my knowledge base here isn't clear enough to be super useful. But since this can't be `Any` I think it is better to explicit:

`str | float | int` or whatever types even if it gets long.

Reply:
You can really put almost anything on the dictionary value. First we need to make a distinction of what we want on it and then find the typing.

I think that you are assuming that we already know what could be there but that's a decision to make. Or maybe I am reading you wrong.

Reply:
Nope, I read your comment wrong. I see now what you meant. I could imagine annotations of things like int, float, string, datetime that people would want to annotate. I guess the question is how much of this annotation machinery would duplicate the goal of what NWB does. I mean neo also has an annotation mechanism, but I've never used it.

Reply:
What I have seen until now is datetime, int, str, float, and nested dictionaries and list of those (for probe properties) . I would wait till something breaks so we can dicuss what should not be there : P

Reply:
Now that seems like a challenge :P

Reply:
Yeah, I mean, if you put a file stream in annotations and then you try to pickle it should break. Or you can put a large numpy arrray  or something of the sort